[BLAS::SYCL-BLAS backend] New BLAS backend: SYCL-BLAS (#262)
oneapi-src/oneMKL: docs/building_the_project.rst

     - | Radeon Instinct(TM) MI 25 Accelerator
       | Radeon(TM) RX Vega 64/56 Graphics

Building for SYCL-BLAS
^^^^^^^^^^^^^^^^^^^^^^

> Note the SYCL-BLAS backend is experimental and only supports a subset of the

Reviewer: Specify these instructions are for linux only (consistent with other backend instructions).

Author: Done in [147e75c](https://github.com/oneapi-src/oneMKL/pull/262/commits/147e75cc9829e1b9f34b50bd4bb73ca427fba02c).

> operations and features for now.

Reviewer: ```suggestion
Note the SYCL-BLAS backend is experimental and currently only supports a subset of the
operations and features.
```

Author: Done in [147e75c](https://github.com/oneapi-src/oneMKL/pull/262/commits/147e75cc9829e1b9f34b50bd4bb73ca427fba02c).

> SYCL-BLAS is used as a header only library.

Reviewer: ```suggestion
The SYCL-BLAS backend uses the `SYCL-BLAS <https://github.com/codeplaysoftware/sycl-blas>` project as a header-only library.
```
In general I think this section should be cleaned up a bit. Is it a requirement that SYCL-BLAS is header-only? Can we use the built library? Are only the commands below the only way to build it or is there some flexibility? I think it would be better to have a section in the SYCL-BLAS readme about building/installing the header-only library (currently this doesn't exist) and just link to that.

Author: Done in [147e75c](https://github.com/oneapi-src/oneMKL/pull/262/commits/147e75cc9829e1b9f34b50bd4bb73ca427fba02c).

Author: Initially we used the built library, but having it as header-only has advantages:
* Avoids compatibility issues between DPC++ versions used to build SYCL-BLAS and OneMKL.
* Avoids issues with inconstent target triples.
* Simplifies device tuning for SYCL-BLAS.
* Overall, it simplifies building OneMKL.

The install folder must be generated via cmake with the following commands:

.. code-block:: bash

   # Inside <path to SYCL-BLAS>
   mkdir build && cd build
   cmake -DCMAKE_INSTALL_PREFIX=<path to SYCL-BLAS install directory> -DINSTALL_HEADER_ONLY=ON ..
   cmake --build . --target install

.. code-block:: bash

   # Inside <path to onemkl>
   mkdir build && cd build
   cmake .. -DENABLE_SYCLBLAS_BACKEND=True                    \
            [-DREF_BLAS_ROOT=<reference_blas_install_prefix>] \ # required only for testing

>             -Dsycl_blas_DIR=<path to SYCL-BLAS install directory>

Reviewer: ```suggestion
            -DSYCLBLAS_DIR=<path to SYCL-BLAS install directory>
```
for consistency, or make the line above `ENABLE_sycl_blas_BACKEND`.

Author: We cannot change `sycl_blas_DIR` to `SYCLBLAS_DIR` since `sycl_blas_DIR` is the name generated by CMake when installing SYCL-BLAS in header-only mode. 

Potentially we could change `ENABLE_SYCLBLAS_BACKEND` to `ENABLE_sycl_blas_BACKEND`. However, this would be inconsistent with other `ENABLE_XXXX_BACKEND` parameters,

Reviewer: Okay, I guess this makes sense the way it is.

     - | Radeon Instinct(TM) MI 25 Accelerator
       | Radeon(TM) RX Vega 64/56 Graphics

Building for SYCL-BLAS
^^^^^^^^^^^^^^^^^^^^^^

Note the SYCL-BLAS backend is experimental and currently only supports a
subset of the operations and features. The SYCL-BLAS backend cannot be enabled
with other backends and can only be used with the compile time dispatch.
The SYCL-BLAS backend uses the `SYCL-BLAS <https://github.com/codeplaysoftware/sycl-blas>`_
project as a header-only library.

* On Linux*

.. code-block:: bash

   # Inside <path to onemkl>
   mkdir build && cd build
   cmake .. -DENABLE_SYCLBLAS_BACKEND=ON \
            -DENABLE_MKLCPU_BACKEND=OFF  \
            -DENABLE_MKLGPU_BACKEND=OFF  \
            -DTARGET_DOMAINS=blas \
            [-DREF_BLAS_ROOT=<reference_blas_install_prefix>] \ # required only for testing
            -Dsycl_blas_DIR=<path to SYCL-BLAS install directory>
   cmake --build .
   ./bin/test_main_blas_ct
   cmake --install . --prefix <path_to_install_dir>

SYCL-BLAS will be downloaded automatically if not found.

> In header only mode SYCL-BLAS can be tuned for specific targets by adding

Reviewer: My understanding is we can only use SYCL-BLAS in header only mode, so I think mentioning the mode might be confusing to users.
```suggestion
SYCL-BLAS can be tuned for specific hardware targets by adding
```

Reviewer 2: Fair, it is updated.

compiler definitions.
The configure step will try to detect automatically which device is used if

> AOT compilation is used via ``-fsycl-targets``.

Reviewer: I think this should be spelled out a bit more for our users. AOT should be "ahead-of-time" but more importantly, how is fsycl-targets set? Is this part of the cmake build system or somehow external to it?

Reviewer 2: This should be clarified now.
